### DISCUSSION TITLE
Audio output with systematic sample loss

### DISCUSSION THREAD

**Discussion by lutzray** [pschatzmann/Adafruit_TinyUSB_Arduino#28] [Q&A]
After calibrating the RP2040 onboard 1 MHz  clock used by `systick`, I can precisely count (±5 ppm) how many samples have been produced by `readCB()`

```
#include "Adafruit_TinyUSB.h"
#include "hardware/structs/systick.h"

Adafruit_USBD_Audio usb;
volatile size_t sample_count = 0;

size_t readCB(uint8_t* data, size_t buff_len, Adafruit_USBD_Audio& ref) {
  int16_t* data16 = (int16_t*)data;
  size_t samples = buff_len / sizeof(int16_t);
  size_t result = 0;

  for (int j = 0; j < samples; j++) {
    sample_count += 1;
    // period should be 200 samples, it is less (196)
    data16[j] = 32000 * sin(6.2831853 * sample_count / 200);
    result += sizeof(int16_t);
  }
  return result;
}

void setup() {
  // Manual begin() is required on core without built-in support e.g. mbed rp2040
  if (!TinyUSBDevice.isInitialized()) {
    TinyUSBDevice.begin(0);
  }
  systick_hw->csr = 0;        //Disable
  systick_hw->rvr = 1000002;  // calibrated with GNSS = 1.000000 second period
  systick_hw->cvr = 0;        //clear the count to force initial reload
  systick_hw->csr = 0x1;      //Enable, with External clock at 1 MHz

  Serial.begin(115200);
  while (!Serial)
    ;  // wait for serial

  // Start USB device as Audio Source
  usb.setReadCallback(readCB);
  usb.begin(44100, 1, 16);

  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}

void loop() {
  if ((systick_hw->csr >> 16) & 1) {
    // print sample_count each 1.000000 second
    // should be 44100 but is 46000, systematically
    Serial.println(sample_count);
    sample_count = 0;
  }
}
```
typical Arduino Serial print out:
```
14:31:42.116 -> 46000
14:31:43.129 -> 46000
14:31:44.135 -> 45954
14:31:45.115 -> 45954
14:31:46.121 -> 46000
```
And a manual count of samples in the recorded `wav` shows coherent discontinuities in the produced sin wave and a shorter period.

Host computer:
```
Model Identifier:	Macmini7,1
Processor Name:	Dual-Core Intel Core i7
System Version:	macOS 12.7.6 (21H1320)
```

**Comment by pratikrath126** [comment]
Sample loss usually means the readCB isn't draining the buffer fast enough or the host's audio pull rate differs. Try increasing the TinyUSB audio buffer size (AUDIO_BUFFER_SAMPLES) or optimizing the callback to copy samples faster. Also verify sample format (mono/stereo, bits) matches both sides. If the host clock drifts slightly, enabling feedback endpoint (freq adjustment) can help. Check that your SYSTICK calibration isn't interfering with interrupt timing. Good luck!

**Reply by lutzray** [reply]
> Also verify sample format (mono/stereo, bits) matches both sides

They do:

```
        TinyUSB Audio:

          Default Input Device: Yes
          Input Channels: 1
          Manufacturer: Seeed
          Current SampleRate: 44100
          Transport: USB
          Input Source: Default

```

> Check that your SYSTICK calibration isn't interfering with interrupt timing

This is done with another program, before running the test.

> optimizing the callback to copy samples faster

Hmm, `readCB()` can't be any simpler...

> Try increasing the TinyUSB audio buffer size (AUDIO_BUFFER_SAMPLES)

Where do I do this?

thanks for your suggestions!

**Reply by lutzray** [reply]
## audio USB is downpitched 2% ##

I'm following up on my own post. Here's another way to state the problem. To reproduce, you need to apply a 1 PPS from any GPS board on pin D10

```
// code illustrating time error with audio USB from pschatzmann's
// Adafruit_TinyUSB_Arduino fork, on a XIAO RP2040 with 1 PPS
// from GPS on pin D10
#include "Adafruit_TinyUSB.h"
Adafruit_USBD_Audio usb;
volatile bool PPSarrived = false;

size_t readCB(uint8_t* data, size_t buff_len, Adafruit_USBD_Audio& ref) {
  int16_t* data16 = (int16_t*)data;
  size_t samples = buff_len / sizeof(int16_t);
  size_t result = 0;
  for (int j = 0; j < samples; j++) {
    if (PPSarrived) {
      PPSarrived = false;
      data16[j] = 32700;
    } else {
      data16[j] = 0;
    }
    result += sizeof(int16_t);
  }
  return result;
}
void PPS_isr(void) {
  PPSarrived = true;
}
void setup() {
  if (!TinyUSBDevice.isInitialized()) {
    TinyUSBDevice.begin(0);
  }
  // GPS 1 PPS pulse on pin D10
  attachInterrupt(D10, PPS_isr, RISING);
  Serial.begin(115200);
  while (!Serial) ;  // wait for serial
  usb.setReadCallback(readCB);
  usb.begin(44100, 1, 16);
  if (TinyUSBDevice.mounted()) {
    TinyUSBDevice.detach();
    delay(10);
    TinyUSBDevice.attach();
  }
}
void loop() {
}
```
The interval between those spike should be 10.00 s, not 10.210 (yes, the framerate of the captured wav is coherent: 44100)

<img width="734" height="376" alt="Screen Shot 2026-02-28 at 9 38 54 PM" src="https://github.com/user-attachments/assets/e208ff72-9bef-4713-99cb-443e9730d5f4" />

I want to use a RP2040 for precise timing but this won't cut it. I did code a working solution on a Teensy4.0 but it is costly...
